ADBDEV-1459 ADBDEV-1468 add behave tests runners

skhomuti opened the pull request:
Now behave tests run with docker-compose gpdb cluster
Add allure report to behave tests
Add some precondition changes for gpdb cluster in docker
See `arenadata/readme.md` for more info

Latest results: https://allure-ee.adsw.io/launch/9619
Please note that some tests have been excluded and will be changed in separate tasks.

skhomuti committed:
ADBDEV-1459 add behave test scripts
Now behave tests runs with docker-compose gpdb cluster
Add allure report to behave tests
Attach gpdb cluster logs to allure report

Maunty commented on arenadata/scripts/run_behave_tests.bash:
```diff
@@ -0,0 +1,40 @@
+
```
- add shebang
- make file executable

skhomuti replied:
done

Maunty commented on arenadata/docker-compose.yaml:
```diff
@@ -0,0 +1,31 @@
+    privileged: true
+    hostname: sdw2
+    volumes:
+      - "$PWD/ssh_keys/id_rsa:/home/gpadmin/.ssh/id_rsa"
```
ssh-keys or ssh_keys ? check your script

skhomuti committed:
ADBDEV-1459 add stdout/stderr attachment

skhomuti committed:
ADBDEV-1459 add init cluster fixture for tests with `concourse_cluster` tag

skhomuti committed:
ADBDEV-1459 split tests to different clusters
fix concourse_cluster fixture. Now cluster start with feature-level scope

skhomuti committed:
ADBDEV-1459 temporary commented concourse_cluster tests
add workaround of too slow chmod

skhomuti committed:
ADBDEV-1459 fixed gpexpand tests

skhomuti committed:
ADBDEV-1459 add sigar install for gpperfmon tests

skhomuti committed:
ADBDEV-1459 prepare sql dump before tests
Move some precondition steps from docker-compose to behave run script
Add sudoers changes for gpadmin passwordless sudo
Skip broken minirepro tests

skhomuti committed:
ADBDEV-1459 exclude cross_subnet and gpssh tests

skhomuti committed:
ADBDEV-1459 add more info to readme

skhomuti commented on gpMgmt/test/behave_utils/arenadata/fixtures.py:
```diff
@@ -0,0 +1,11 @@
+from behave import fixture
```
This file currently not used but will need for `concourse_cluster` tests later

skhomuti commented on gpMgmt/test/behave/mgmt_utils/minirepro.feature:
```diff
@@ -1,4 +1,4 @@
-@minirepro
+@minirepro @skip
```
Skip because these tests are broken

skhomuti commented on gpMgmt/Makefile.behave:
```diff
@@ -1,14 +1,15 @@
+		PYTHONPATH=$$PYTHONPATH:$(PEXPECT_LIB):$(TEST_DIR) behave $(CURDIR)/test/behave/* -s -k $(flags) 2>&1 ; \
 	elif [ -n """$(tags)""" ]; then \
-		PYTHONPATH=$$PYTHONPATH:$(PEXPECT_LIB) behave $(CURDIR)/test/behave/* -s -k --tags=$(tags) 2>&1 ; \
+		PYTHONPATH=$$PYTHONPATH:$(PEXPECT_LIB):$(TEST_DIR) behave $(CURDIR)/test/behave/* -s -k --tags=$(tags) 2>&1 ; \
```
Changes for custom formatter support

skhomuti commented on concourse/scripts/setup_gpadmin_user.bash:
```diff
@@ -41,7 +41,8 @@ transfer_ownership() {
     [ -d /usr/local/greenplum-db-devel ] && chown -R gpadmin:gpadmin /usr/local/greenplum-db-devel
-    chown -R gpadmin:gpadmin /home/gpadmin
+    # workaround of too slow recursive chown https://github.com/docker/for-linux/issues/388
+    find /home/gpadmin -not -group gpadmin -not -user gpadmin -print0 | xargs -P 0 -0 --no-run-if-empty chown --no-dereference gpadmin:gpadmin
```
It's dumb but it works and speed-up chown

dgusakov commented on arenadata/readme.md:
```diff
@@ -0,0 +1,30 @@
+Behave tests now can run locally with docker-compose.
+
+Feature files are located in `gpMgmt/test/behave/mgmt_utils`
+Before run tests you need docker-image
```
```suggestion
Before run tests you need to build a docker-image
```

dgusakov commented on arenadata/readme.md:
````diff
@@ -0,0 +1,30 @@
+```
+
+
+Tests are used `allure-behave` package and store allure output files in `allure-results` folder
````
```suggestion
Tests use `allure-behave` package and store allure output files in `allure-results` folder
```

dgusakov commented on arenadata/readme.md:
```diff
@@ -0,0 +1,30 @@
+
+
+Tests are used `allure-behave` package and store allure output files in `allure-results` folder
+Pay attention that `allure-behave` has too old a version because it is compatible with `python2`.
```
```suggestion
**NOTE** that `allure-behave` has too old a version because it is compatible with `python2`.
```

dgusakov commented on arenadata/readme.md:
```diff
@@ -0,0 +1,30 @@
+
+Tests in a docker-compose cluster use the same ssh keys for `gpadmin` user and pre-add the cluster hosts to `.ssh/know_hosts` and `/etc/hosts`.
+
+Docker containers have installed `sigar` libraries. It required only for `gpperfmon` tests.
```
```suggestion
Docker containers have installed `sigar` libraries. It is required only for `gpperfmon` tests.
```

skhomuti committed:
ADBDEV-1459 readme fixes

Stolb27 commented on arenadata/scripts/behave_gpdb.bash:
```diff
@@ -0,0 +1,65 @@
+
+    pip --retries 10 install --ignore-installed pip allure-behave==2.4.0
+    rpm -i https://ci.arenadata.io/artifactory/ADB/6.7.1_arenadata4/centos/7/community/x86_64/sigar-1.6.5-163.el7.x86_64.rpm &&
+    rpm -i https://ci.arenadata.io/artifactory/ADB/6.7.1_arenadata4/centos/7/community/x86_64/sigar-headers-1.6.5-163.el7.x86_64.rpm &&
```
Maybe modify the Docker image to persist `sigar` to avoid duplication of these lines? I don't think that it's wanted behavior for regression tests.

skhomuti replied:
fixed

Stolb27 commented on arenadata/scripts/behave_gpdb.bash:
```diff
@@ -0,0 +1,65 @@
+    gpstop -u
+    mkdir -p /home/gpadmin/sqldump
+    time /home/gpadmin/gpdb_src/concourse/scripts/dumpdb.bash
+    xz -d /home/gpadmin/sqldump/dump.sql.xz
```
Why we dump an empty database (cluster just created with `make_cluster`)? gpexpand tests expect complex database schema [produced](https://github.com/arenadata/gpdb/blob/3d6cc3d62b2b800bf72afa0d4a35333f4c543f00/concourse/scripts/ic_gpdb.bash#L80) by regression run.

Stolb27 replied:
I think we may either spaces or tabs in this function)

skhomuti replied:
https://arenadata.atlassian.net/browse/ADBDEV-1901 I suggest making the correct dump a separate task

Maunty commented on arenadata/docker-compose.yaml:
```diff
@@ -0,0 +1,62 @@
+    privileged: true
+    entrypoint: >
+      bash -c "mkdir -p /data/gpdata && chmod -R 777 /data &&
+      sysctl 'kernel.sem=500 1024000 200 4096' &&
```
https://docs.docker.com/engine/reference/commandline/run/#configure-namespaced-kernel-parameters-sysctls-at-runtime

dgusakov replied:
@skhomuti take a look

skhomuti replied:
fix

skhomuti committed:
ADBDEV-1459 sysctl as a docker-compose properties

skhomuti committed:
ADBDEV-1459 move sigar packages to base image stage

skhomuti committed:
ADBDEV-1459 removed useless volumes mount

Stolb27 commented on arenadata/docker-compose.yaml:
```diff
@@ -0,0 +1,64 @@
+version: "3"
+services:
+  mdw:
+    image: "hub.adsw.io/library/gpdb_regress:${BRANCH_NAME}"
```
image name has been renamed to gpdb6_regress since adb-5.x tests were implemented. Please check it.

skhomuti committed:
ADBDEV-1459 fixed tabs and spaces
fix docker image name
revert recursive chown

skhomuti committed:
ADBDEV-1459 fix requirements resolving
